src/storage/DataAccessorBasedStore.ts

```diff
@@ -0,0 +1,366 @@
+        get data(): Readable {
+          // This allows other modules to still add metadata before the output data is written
+          return streamifyArray(result.metadata.quads());
+        },
+        metadata,
+      };
+    } else {
+      // Data resource
+      result = {
+        binary: metadata.contentType !== INTERNAL_QUADS,
+        data: await this.accessor.getData(identifier),
+        metadata,
+      };
+    }
+    return result;
+  }
+
+  public async addResource(container: ResourceIdentifier, representation: Representation): Promise<ResourceIdentifier> {
+    this.validateIdentifier(container);
+    await this.accessor.canHandle(representation);
+
+    // Using the parent metadata as we can also use that later to check if the nested containers maybe need to be made
+    const parentMetadata = await this.getSafeNormalizedMetadata(container);
+
+    // When a POST method request targets a non-container resource without an existing representation,
+    // the server MUST respond with the 404 status code.
+    if (!parentMetadata && !container.path.endsWith('/')) {
+      throw new NotFoundHttpError();
+    }
+
+    const newID = this.createSafeUri(container, representation.metadata, parentMetadata);
+
+    if (parentMetadata && !this.isExistingContainer(parentMetadata)) {
+      throw new MethodNotAllowedHttpError('The given path is not a valid container.');
+    }
+
+    // Need at least 1 container if parent doesn't exist yet
+    await this.writeData(newID, representation, newID.path.endsWith('/'), !parentMetadata);
+
+    return newID;
+  }
+
+  public async setRepresentation(identifier: ResourceIdentifier, representation: Representation): Promise<void> {
+    this.validateIdentifier(identifier);
+    await this.accessor.canHandle(representation);
+
+    // Check if the resource already exists
+    const oldMetadata = await this.getSafeNormalizedMetadata(identifier);
+
+    // Might want to redirect in the future
+    if (oldMetadata && oldMetadata.identifier.value !== identifier.path) {
+      throw new ConflictHttpError(`${identifier.path} conflicts with existing path ${oldMetadata.identifier.value}`);
+    }
+
+    // If we already have a resource for the given identifier, make sure they match resource types
+    const isContainer = this.isNewContainer(representation.metadata, identifier.path);
+    if (oldMetadata && isContainer !== this.isExistingContainer(oldMetadata)) {
+      throw new ConflictHttpError('Input resource type does not match existing resource type.');
+    }
+    if (isContainer !== identifier.path.endsWith('/')) {
+      throw new UnsupportedHttpError('Containers should have a `/` at the end of their path, resources should not.');
+    }
+
+    // Potentially have to create containers if it didn't exist yet
+    await this.writeData(identifier, representation, isContainer, !oldMetadata);
+  }
+
+  public async modifyResource(): Promise<void> {
+    throw new Error('Not supported.');
```

Comment:
We probably want an error type for that: https://tools.ietf.org/html/rfc7231#section-6.6.2